ops/pebble.py

```diff
@@ -1579,28 +1573,47 @@ def _make_auth_dict(permissions, user_id, user, group_id, group) -> typing.Dict:
             d['group'] = group
         return d
 
-    @staticmethod
-    def _encode_multipart(metadata, path, content):
+    def _encode_multipart(self, metadata, path, source, encoding):
         # Python's stdlib mime/multipart handling is screwy and doesn't handle
         # binary properly, so roll our own.
+
+        if isinstance(source, str):
+            source = io.StringIO(source)
+        elif isinstance(source, bytes):
+            source = io.BytesIO(source)
+
         boundary = binascii.hexlify(os.urandom(16))
         path_escaped = path.replace('"', '\\"').encode('utf-8')  # NOQA: test_quote_backslashes
-        parts = []
-        parts.extend([
-            b'--', boundary, b'\r\n',
-            b'Content-Type: application/json\r\n',
-            b'Content-Disposition: form-data; name="request"\r\n',
-            b'\r\n',
-            json.dumps(metadata).encode('utf-8'), b'\r\n',
-            b'--', boundary, b'\r\n',
-            b'Content-Type: application/octet-stream\r\n',
-            b'Content-Disposition: form-data; name="files"; filename="', path_escaped, b'"\r\n',
-            b'\r\n',
-            content, b'\r\n',
-            b'--', boundary, b'--\r\n',
-        ])
         content_type = 'multipart/form-data; boundary="' + boundary.decode('utf-8') + '"'
-        return b''.join(parts), content_type
+
+        def generator():
+            nonlocal source
```

Comment:
Nice use of a generator! I didn't know urllib.Request could take a generator -- that's great.

Side note: `source` doesn't need to be a `nonlocal` -- you only need `nonlocal` if you're assigning to a variable (`nonlocal` and `global` have to do with where a name is bound / assigned). Just referencing it is fine without any `nonlocal`.

Reply:
I think I had assigned to source in a previous version while developing, added the keyword out of necessity at the time, then didn't check if it was still needed.  I can remove it; it looks like it's no longer needed.

Reply:
I like the generator, too. Nicely Pythonic, and an elegant way of doing things.